pandas/core/internals/managers.py

```diff
@@ -1233,6 +1233,12 @@ def value_getitem(placement):
             blk = self.blocks[blkno]
             blk_locs = blklocs[val_locs.indexer]
             if blk.should_store(value):
+                if (
+                    value.shape[0] != 1
```

Comment:
probably better to have `.should_store` return False in this case and have it hit the other path? or is that incorreect

Reply:
If I understand you correctly, I think it should stay as it is because the below test case fails:
```python
def test_setitem_size_incompatible_ndarray2(arr):
    data = DataFrame(
        [[1, "A", 1.0], [2, "B", 2.0], [3, "C", 3.0], [4, "D", 4.0]],
        columns=["A", "A", "B"],
    )
    msg = "Errored123"
    with pytest.raises(ValueError, match=msg):
        # data is of shape (4,2) but we are assigning it to shape (4,4)
        data["A"] = np.random.randn(4, 4)
```
It fails because it goes down the other path (where `blk.should_store` is false) and no error is thrown despite assigning the column(s) to data of a different shape. I haven't handled this yet because I wanted to make sure the logic was ok so far (sorry I should have mentioned this earlier).

Reply:
I think that `.should_store` is probably not the best description of what it does, (based on my understanding of the method).